### PR TITLE
feat(@vtmn/react): add divider component

### PR DESCRIPTION
**Pull request by 0x2A-git**
## Pull Request checklist

🚨 Please review the [guidelines for contributing](../blob/main/CONTRIBUTING.md) to this repository.

- [x] Make sure you are requesting to **pull a topic/feature/bugfix branch**. Don't request your main!
- [x] Check the commit's or even all commits' message styles matches our requested structure.
- [x] Check your code additions will fail neither code linting checks nor unit test.
- [x] Check PR name, it must follow the https://www.conventionalcommits.org pattern
- [x] If it's a new component in CSS, ask for design review

## Description

This P.R aims to add a Divider component for vtmn/react ( related issue : #813 ).

## Does this introduce a breaking change?

- No

## Other information

Concerns : 

1 -  <del>When using the text prop to indicate where the label should go it seems to work the other way around, IE: if I set text to 'end', the text goes to the left ( is it the intended behavior or am I missing something ? )</del>

2 - <del>It is my first time using StoryBook, so I don't know if the story is good enough because when using text='center' it just doesn't show the divider's bar when there is text because it must be in a container such as a flexbox in order to work. Since the other components of the project weren't encapsulated in divs I didn't encapsulated Divider in a div. </del>

3 - <del>I'm not sure if the figma link I put will work ( When trying to access [Divider Figma Page](https://www.figma.com/file/zDZIyayUlr1yTWrsi7cFoo/Vitamin---Web-UI-kit-(v0)?node-id=6871%3A13107) it doesn't seem to work ? ) </del>

Suggestions : 

1 - <del>Shouldn't the text prop be equals to 'center' by default, I don't really see the point of setting it to null ? ( Did it anyway because I've seen it in the props section [here](https://www.decathlon.design/726f8c765/p/28ad9b-divider-beta/t/809c31) )</del>

❤️ Thank you!

### CODE REVIEW

**Comment by lauthieb** [comment]
Thanks a lot @0x2A-git for this contribution!
I'll check soon :)

**Comment by 0x2A-git** [comment]
Okay, so I've just found out about the CSS Storybook 🥴 I'm going to try to replicate the story in React and apply some fixes to the component :sweat_smile:

File: packages/showcases/core/csf/components/VtmnDivider.csf.js [hunk 1]
```diff
@@ -0,0 +1,36 @@
+export const argTypes = {
+  orientation: {
+    type: { name: 'string', required: false },
+    description: 'The orientation of the divider.',
+    defaultValue: 'horizontal',
+    control: {
+      type: 'text',
+    },
+  },
+  text: {
```
Comment (lauthieb) [inline packages/showcases/core/csf/components/VtmnDivider.csf.js:10]:
```suggestion
  textPosition: {
```

File: packages/sources/react/src/components/VtmnDivider/VtmnDivider.tsx [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+   * The orientation of the divider
+   * @defaultValue 'horizontal'
+   */
+  orientation?: VtmnDividerOrientation;
+
+  /**
+   * The position of the text inside the divider
+   * @defaultValue null
+   */
+  text?: VtmnDividerText;
```
Comment (lauthieb) [inline packages/sources/react/src/components/VtmnDivider/VtmnDivider.tsx:18]:
```suggestion
  textPosition?: VtmnDividerText;
```

File: packages/sources/react/src/components/VtmnDivider/VtmnDivider.tsx [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+  extends React.ComponentPropsWithoutRef<'div'> {
+  /**
+   * The orientation of the divider
+   * @defaultValue 'horizontal'
+   */
+  orientation?: VtmnDividerOrientation;
+
+  /**
+   * The position of the text inside the divider
+   * @defaultValue null
```
Comment (lauthieb) [inline packages/sources/react/src/components/VtmnDivider/VtmnDivider.tsx:16]:
```suggestion
   * @defaultValue undefined
```

File: packages/showcases/core/csf/components/VtmnDivider.csf.js [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    type: { name: 'string', required: false },
+    description: 'The position of the text inside the divider.',
+    defaultValue: null,
```
Comment (lauthieb) [inline packages/showcases/core/csf/components/VtmnDivider.csf.js:13]:
```suggestion
    defaultValue: 'start',
```

File: packages/sources/react/src/components/VtmnDivider/VtmnDivider.tsx [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+   */
+  text?: VtmnDividerText;
+}
+
+export const VtmnDivider = ({
+  orientation = 'horizontal',
+  text = null,
```
Comment (lauthieb) [inline packages/sources/react/src/components/VtmnDivider/VtmnDivider.tsx:23]:
```suggestion
  textPosition = 'start',
```

File: packages/sources/react/src/components/VtmnDivider/types.ts [hunk 6]
```diff
@@ -0,0 +1,2 @@
+export type VtmnDividerText = 'start' | 'center' | 'end' | null;
```
Comment (lauthieb) [inline packages/sources/react/src/components/VtmnDivider/types.ts:1]:
```suggestion
export type VtmnDividerText = 'start' | 'center' | 'end';
```

File: packages/sources/react/src/components/VtmnDivider/VtmnDivider.tsx [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+  children,
+  className,
+  ...props
+}: VtmnDividerProps) => {
+  return (
+    <div
+      className={clsx(
+        'vtmn-divider',
+        `vtmn-divider_text--${text == null ? 'center' : text}`,
```
Comment (lauthieb) [inline packages/sources/react/src/components/VtmnDivider/VtmnDivider.tsx:32]: You need to apply this CSS class only if children is defined.
Can you please try this below?
```suggestion
        children && `vtmn-divider_text-position--${textPosition}`,
```

File: packages/sources/react/src/components/VtmnDivider/VtmnDivider.tsx (hunk 3 above)
Comment (lauthieb) [inline packages/sources/react/src/components/VtmnDivider/VtmnDivider.tsx:16]:
```suggestion
   * @defaultValue 'start'
```

File: packages/showcases/react/stories/components/VtmnDivider/VtmnDivider.stories.tsx [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+      </VtmnDivider>
+      <p>
+        Lorem ipsum dolor sit amet consectetur adipisicing elit. Magni quidem
+        architecto cum assumenda ad possimus modi harum voluptas earum in?
+      </p>
+    </div>
+  </div>
+);
+
+export const Vertical = VerticalTemplate.bind({});
```
Comment (lauthieb) [inline packages/showcases/react/stories/components/VtmnDivider/VtmnDivider.stories.tsx:162]: We need to use the same behavior as other React storybook stories like `VtmnButton`, `VtmnToggle` etc. :)

```suggestion
import React from 'react';
import { VtmnButton } from '@vtmn/react';
import { Meta, Story } from '@storybook/react';
import {
  argTypes,
  parameters,
} from '@vtmn/showcase-core/csf/components/VtmnDivider.csf';

export default {
  title: 'Components/VtmnDivider',
  component: VtmnDivider,
  argTypes: {
    ...argTypes,
    children: {
      type: { name: 'string', required: false },
      description: 'The content to render inside the component',
      defaultValue: 'Button',
      control: {
        type: 'text',
      },
    },
  },
  parameters,
} as Meta;

const Template: Story = (args) => <VtmnDivider {...args} />;

export const Overview = Template.bind({});
```

**Review by lauthieb — CHANGES_REQUESTED** [review]
Thanks a lot for your contribution!

As we are in beta phase for divider, we have changed some different things in specifications of the component:

Capture d’écran 2021-12-31 à 17 15 36

That's why I've proposed you some changes!

Feel free if you have question :-)

**Comment by lauthieb** [comment]
> 3 - ~I'm not sure if the figma link I put will work ( When trying to access [Divider Figma Page](https://www.figma.com/file/zDZIyayUlr1yTWrsi7cFoo/Vitamin---Web-UI-kit-(v0)?node-id=6871%3A13107) it doesn't seem to work ? ) ~

Yes but this Figma is private for now sorry, but I confirm that it works for Decathlon people :)

> Suggestions :
> 
> 1 - ~Shouldn't the text prop be equals to 'center' by default, I don't really see the point of setting it to null ? ( Did it anyway because I've seen it in the props section [here](https://www.decathlon.design/726f8c765/p/28ad9b-divider-beta/t/809c31) )~

Yes you're right, we will now have 'start' by default and "text" has been renamed to "textPosition", you will see that in my different propositions of code changes :)

Thanks!

File: packages/showcases/react/stories/components/VtmnDivider/VtmnDivider.stories.tsx (hunk 8 above)
Comment (0x2A-git) [reply packages/showcases/react/stories/components/VtmnDivider/VtmnDivider.stories.tsx:162]: Thanks for the review @lauthieb  ! I may have a question about the storybook part, in fact the divider doesn't render as it should by calling <VtmnDivider {...args} /> without a container that wraps it, this is mainly why I've done examples with text, fleboxes and grids ->

Without container :
Screenshot 2021-12-31 at 18-05-28 Components VtmnDivider - Overview ⋅ Storybook

With container (200x200 pixels) wrapping <VtmnDivider {...args} />
Screenshot 2021-12-31 at 18-06-06 Components VtmnDivider - Overview ⋅ Storybook

There are also some ( strange ? ) behaviors that I forgot to mention, for example when there isn't enough content I think the divider works the other way around with the orientation set to vertical  : 

Screenshot 2021-12-31 at 18-13-50 Components VtmnDivider - Overview ⋅ Storybook

Shouldn't textPosition=end place text at the bottom with most of the line above the label and the rest under the label like here ?

image


These weird placements do not occur with enough text. From what I understood it occurs when the height of the divider isn't tall enough.

**Comment by lauthieb** [comment]
Hi @0x2A-git thanks a lot for your answer. Is it possible to meet you in videoconference / call soon in order to speak about your 2 Pull Requests and pair programming together?
You can contact me in our slack: https://join.slack.com/t/decathlon-design/shared_invite/zt-ou0n9qas-n_oamDSVUIqvLqNO1LETJg 
Thanks a lot!
Laurent

**Comment by lauthieb** [comment]
Hello @0x2A-git, as seen together, we will meet each other in one week. For the moment, I convert this PR to draft :)

**Review by lauthieb — APPROVED** [review]
Thanks a lot @0x2A-git for this great contribution to Decathlon Design System, you rock 🚀

**Pull request merged**